gobblin-utility/src/main/java/org/apache/gobblin/fsm/FiniteStateMachine.java

```diff
@@ -0,0 +1,357 @@
+	private final Condition condition;
+	private final T initialState;
+
+	private volatile T currentState;
+	private volatile Transition currentTransition;
+
+	private FiniteStateMachine(SetMultimap<T, T> allowedTransitions, Set<T> universalEnds, T initialState) {
+		this.allowedTransitions = allowedTransitions;
+		this.universalEnds = universalEnds;
+
+		this.lock = new ReentrantLock();
+		this.condition = this.lock.newCondition();
+		this.initialState = initialState;
+		this.currentState = initialState;
+	}
+
+	/**
+	 * Start a transition to the end state specified. The returned {@link Transition} object is a closeable that will finalize
+	 * the transition when it is closed. While the transition is open, no other transition can start.
+	 *
+	 * It is recommended to call this method only within a try-with-resource block to ensure the transition is closed.
+	 *
+	 * @throws UnallowedTransitionException If the transition is not allowed.
+	 * @throws InterruptedException if the thread got interrupted while waiting for a non-transitioning state.
+	 */
+	public Transition startTransition(T endState) throws UnallowedTransitionException, InterruptedException {
+		try {
+			this.lock.lock();
+			while (isTransitioning()) {
+				this.condition.await(1, TimeUnit.SECONDS);
+			}
+			if (!isAllowedTransition(this.currentState, endState)) {
+				throw new UnallowedTransitionException(this.currentState, endState);
+			}
+			Transition transition = new Transition(endState);
+			this.currentTransition = transition;
+			return transition;
+		} finally {
+			this.lock.unlock();
+		}
+	}
+
+	/**
+	 * Transition immediately to the given end state. This is essentially {@link #startTransition(Object)} immediately
+	 * followed by {@link Transition#close()}.
+	 *
+	 * @throws UnallowedTransitionException if the transition is not allowed.
+	 * @throws InterruptedException if the thread got interrupted while waiting for a non-transitioning state.
+	 */
+	public void transitionImmediately(T endState) throws UnallowedTransitionException, InterruptedException {
+		Transition transition = startTransition(endState);
+		transition.close();
+	}
+
+	/**
+	 * Transition immediately to the given end state if the transition is allowed.
+	 *
+	 * @return true if the transition happened.
+	 * @throws InterruptedException if the thread got interrupted while waiting for a non-transitioning state.
+	 */
+	public boolean transitionIfAllowed(T endState) throws InterruptedException {
+		try {
+			transitionImmediately(endState);
+		} catch (UnallowedTransitionException exc) {
+			return false;
+		}
+		return true;
+	}
+
+	/**
+	 * Get the current state. This method will wait until the FSM is in a non-transitioning state (although a transition
+	 * may start immediately after).
+	 * @throws InterruptedException if the thread got interrupted while waiting for a non-transitioning state.
+	 */
+	public T getCurrentState() throws InterruptedException {
+		try {
+			this.lock.lock();
```

Comment:
Why do we need a lock for getting current state? If we do indeed need a lock, can we have a ReadWriteLock that will allow multiple concurrent readers?